File: Data Structure/Trie/LET_no208_Implement Trie (Prefix Tree).cpp
class Trie {
private:
    unordered_map<char, Trie*> children;
    bool is_word;
public:
    Trie() { is_word = false; }
    
    void insert(string word) {
        Trie* cur = this;
        for (char c : word) {
            if (!cur->children.count(c))
                cur->children[c] = new Trie();
            cur = cur->children[c];
        }
        cur->is_word = true;
    }
    
    bool search(string word) {
        Trie* cur = this;
        for (char c : word) {
            if (!cur->children.count(c))
                return false;
            cur = cur->children[c];
        }
        return cur->is_word;
    }
    
    bool startsWith(string prefix) {
        Trie* cur = this;
        for (char c : prefix) {
            if (!cur->children.count(c))
                return false;
            cur = cur->children[c];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: Data Structure/Trie/LET_no208_Implement Trie (Prefix Tree).py
class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        cur = self.root
        for c in word:
            cur = cur.children[c]
        cur.is_word = True

    def search(self, word: str) -> bool:
        cur = self.root
        for c in word:
            cur = cur.children.get(c)
            if not cur:
                return False
        return cur.is_word

    def startsWith(self, prefix: str) -> bool:
        cur = self.root
        for c in prefix:
            cur = cur.children.get(c)
            if not cur:
                return False
        return True
        
class TrieNode:
    def __init__(self):
        self.children = collections.defaultdict(TrieNode)
        self.is_word = False

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)